src/http/headers.cr

```diff
@@ -55,37 +63,73 @@ struct HTTP::Headers
+  # `ArgumentError` will be raised if header includes invalid characters.
+  #
+  # ```
+  # headers = HTTP::Headers{"host" => "crystal-lang.org"}
+  # headers["host"] = "play.crystal-lang.org"
+  # headers # => HTTP::Headers{"host" => "play.crystal-lang.org"}
+  # headers["accept-encoding"] = "text/html"
+  # headers # => HTTP::Headers{"host" => "play.crystal-lang.org", "accept-encoding" => "text/html"}
+  # ```
+  def []=(key : String, value : String) : String
     check_invalid_header_content(value)
 
     @hash[wrap(key)] = value
   end
 
-  def []=(key, value : Array(String))
+  # Sets the value of key to the given value.
+  # `ArgumentError` will be raised if header includes invalid characters.
+  #
+  # ```
+  # headers = HTTP::Headers{"host" => "crystal-lang.org"}
+  # headers["host"] = ["crystal-lang.org", "play.crystal-lang.org"]
+  # headers # => HTTP::Headers{"host" => ["crystal-lang.org", "play.crystal-lang.org"]}
+  # headers["accept-encoding"] = ["text/html", "application/json"]
+  # headers # => HTTP::Headers{"host" => ["crystal-lang.org", "play.crystal-lang.org"], "accept-encoding" => ["text/html", "application/json"]}
+  # ```
+  def []=(key : String, value : Array(String)) : Array(String)
     value.each { |val| check_invalid_header_content val }
 
     @hash[wrap(key)] = value
   end
 
-  def [](key)
+  # Returns the value for the key given by key.  If no value found a `KeyError` will be raised.
+  # `ArgumentError` will be raised if header includes invalid characters.
+  #
+  # ```
+  # headers = HTTP::Headers{"host" => "crystal-lang.org", "accept-encoding" => ["text/html", "application/json"]}
+  # headers["host"]            # => "crystal-lang.org"
+  # headers["accept-encoding"] # => "text/html,application/json"
+  # headers["user-agent"]      # => raises KeyError
+  # ```
+  def [](key : HTTP::Headers::Key | String) : String
     values = @hash[wrap(key)]
     concat values
   end
 
-  def []?(key)
-    fetch(key, nil)
+  # Returns the value for the key given by key. If not found, returns `nil`.
+  # `ArgumentError` will be raised if header includes invalid characters.
+  #
+  # ```
+  # headers = HTTP::Headers{"host" => "crystal-lang.org", "accept-encoding" => ["text/html", "application/json"]}
+  # headers["host"]?            # => "crystal-lang.org"
+  # headers["accept-encoding"]? # => "text/html,application/json"
+  # headers["user-agent"]?      # => nil
+  # ```
+  def []?(key : HTTP::Headers::Key | String) : String?
+    fetch(key, nil).as(String?) # TODO: fixup
   end
 
   # Returns if among the headers for *key* there is some that contains *word* as a value.
   # The *word* is expected to match between word boundaries (i.e. non-alphanumeric chars).
   #
   # ```
-  # require "http/headers"
-  #
-  # headers = HTTP::Headers{"Connection" => "keep-alive, Upgrade"}
-  # headers.includes_word?("Connection", "Upgrade") # => true
+  # headers = HTTP::Headers{"accept-encoding" => "text/html,application/json"}
```

Comment:
`require "http/headers"` should be left and added to the other examples. We want reproducible code examples.